karma/module.py

```diff
@@ -0,0 +1,641 @@
+import asyncio
+import contextlib
+from typing import Optional, List, Tuple
+
+from emoji import UNICODE_EMOJI as _UNICODE_EMOJI
+
+
+import discord
+from discord.ext import commands, tasks
+
+from core import check, i18n, logger, utils
+from core import TranslationContext
+
+from .database import (
+    KarmaMember,
+    KarmaEmoji,
+    UnicodeEmoji,
+    DiscordEmoji,
+    IgnoredChannel,
+    BoardOrder,
+    BoardType,
+)
+
+UNICODE_EMOJI = _UNICODE_EMOJI["en"]
+del _UNICODE_EMOJI
+
+_ = i18n.Translator("modules/boards").translate
+bot_log = logger.Bot.logger()
+guild_log = logger.Guild.logger()
+
+
+class Karma(commands.Cog):
+    def __init__(self, bot):
+        self.bot = bot
+
+        self.value_cache = {}
+        self.given_cache = {}
+        self.taken_cache = {}
+
+        self.karma_cache_write.start()
+
+    @tasks.loop(seconds=30.0)
+    async def karma_cache_loop(self) -> None:
```

Comment:
It would be good to call this function on module unload (or/and bot shutdown), 'cause now if you unload module or shutdown bot, some data could be lost.

Reply:
We can just use before and after loops shown in the discord.py documentation [here](https://discordpy.readthedocs.io/en/stable/ext/tasks/index.html).

```py
async def do_bulk(self):
        # bulk insert data here
        ...

    @tasks.loop(seconds=10.0)
    async def bulker(self):
        async with self.lock:
            await self.do_bulk()
    
    @bulker.before_loop
    async def before_bulker(self):
        # Waiting until the bot is ready before the loop starts
        print('waiting...')
        await self.bot.wait_until_ready()

    @bulker.after_loop
    async def on_bulker_cancel(self):
        if self.bulker.is_being_cancelled() and len(self._batch) != 0:
            # if we're cancelled and we have some data left...
            # let's insert it to our database
            await self.do_bulk()
```